Fixes #2118 Add a Publications Content Type to QS2

tadean opened the pull request:
<!--- Provide a general summary of your changes in the Title above --><!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail (include keywords close/fix/resolve) -->
This PR adds a Pupblication content type to Quickstart. It contains a Publication content type, an Author entity that is used for representing publication authors (with optional referential linking to a Quickstart Person). It presently contains styles for MLA, IEEE, AMA, Bluebook Law Review, APA, and Chicago styles. It uses [Citation Style Langauge](https://citationstyles.org/) for bibliographic rendering, via a PHP CSL 1.0.1 processor ([seboettg/citeproc-php](https://github.com/seboettg/citeproc-php))

- Additional views to be added
- Role permissions to be added

## Related issues
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->
#2118 

## How to test

- Enable **Quickstart Publication (Experimental)**
- Create a flexible page that adds displays of the AZ Publications views to the content
- The AZ Publication views currently have one display per citation style standard
- Create publication content
- View the publication output on your view page for correctness
- Not all fields are used by every publication type and style standard. Depending on the type of publication and citation style some fields may not be displayed if the style does not use them.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

### Arizona Quickstart (install profile, custom modules, custom theme)
- **Patch release changes**
   - [ ] Bug fix
   - [ ] Accessibility, performance, or security improvement
   - [ ] Critical institutional link or brand change
   - [x] Adding experimental module
   - [ ] Update experimental module
- **Minor release changes**
   - [x] New feature
   - [ ] Breaking or visual change to existing behavior
   - [ ] Upgrade experimental module to stable
   - [ ] Enable existing module by default or database update
   - [ ] Non-critical brand change
   - [x] New internal API or API improvement with backwards compatibility
   - [ ] Risky or disruptive cleanup to comply with coding standards
   - [ ] High-risk or disruptive change (requires upgrade path, risks regression, etc.)
- **Other or unknown**
   - [ ] Other or unknown

### Drupal core
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch level release (non-security bug-fix release)
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major or minor level update
- **Other or unknown**
   - [ ] Other or unknown

### Drupal contrib projects
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch or minor level update
   - [x] Add new module
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major level update
- **Other or unknown**
   - [ ] Other or unknown

## Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

tadean commented:
**Todo:**

- [ ] Fix phpstan entity warnings
- [ ] Add permissions (pending discussion)
- [ ] Potentially move stylesheets into configuration entities (pending discussion)
- [ ] Render additional links in citation
- [ ] Add additional views

UAScotttJaxon commented:
Can we include PubMed ID in this module? College of Medicine/UAHS sites utilize PubMed ID. https://www.drupal.org/project/bibcite_pubmed

joshuasosa reviewed (commented):
Are all these locale files necessary? It seems like AZQS is primarily focused on English and doesn't provide any other translations for anything else. Could citation-style-language/locales be reduced to mostly just EN to contain the scope?

tadean commented:
> Are all these locale files necessary? It seems like AZQS is primarily focused on English and doesn't provide any other translations for anything else. Could citation-style-language/locales be reduced to mostly just EN to contain the scope?

I agree - I'm hoping to make a change that moves the locale files to the vendor folder (via a CSL metapackage) and moves the stylesheets to configuration entities so other stylesheets can be added to the site without needing to change the module.

tadean commented:
> Can we include PubMed ID in this module? College of Medicine/UAHS sites utilize PubMed ID. https://www.drupal.org/project/bibcite_pubmed

We discussed adding something for this today and it seemed reasonable. The one caveat is that not all style standards have a a rule for displaying a pubmed (or for example, a DOI) link, so whether that ID is used would depend on the style standard.

tadean commented:
Todo:
- [ ] Move stylesheets into configuration entities
- [ ] Investigate moving locales to `/vendor`
- [ ] Add additional views
- [ ] Investigate pubmed/DOI ID

tadean commented:
followup research note: PubMed IDs appear as a custom variable in the National Library of Medicine style and several others based upon it. It is typically exposed to CSL processing as the variable `PMID`

DOI numbers are part of the CSL standard schema and use the identifier `DOI`.

tadean commented:
- Removed CSL stylesheets as hardcoded module files in favor of a new Citation Style configuration entity type. This moves the stylesheets into configuration and makes it possible to add additional styles to a site without editing the module. Administration is at `/admin/structure/az_citation_style`
- Removed locale files from the module in favor of using `academicpuma/locales` and adding a patch to allow the CSL processor to locate and use them. This greatly reduces the size of the module but makes it more difficult to change the locale files if needed. However, the locale files change **much** less frequently than the stylesheets and I propose we can make any occasional modifications to the locale files that are needed via composer patches.

tadean commented:
Updated the drupal views patch for the rendered content issue to be compatible with 9.5.x.

bberndt-uaz commented on modules/custom/az_publication/az_publication.module:
```diff
@@ -0,0 +1,448 @@
+      $entity->hasField('field_az_publication_type') &&
+      !empty($entity->field_az_publication_type->value)) {
+    if ($display->getComponent('az_publication_bibliography')) {
+      // Publicaiton fields to CSL schema mappings.
```
```suggestion
      // Publication fields to CSL schema mappings.
```

bberndt-uaz commented on modules/custom/az_publication/az_publication.module:
```diff
@@ -0,0 +1,448 @@
+ * into CSL format. This is an object with a date-parts property and the primary
+ * processing needing to be done is making sure the date-parts nested array has
+ * the number of components dictated by the date type, e.g. we do not want to
+ * send a date with three componens if there should be only a year component.
```
```suggestion
 * send a date with three components if there should be only a year component.
```

joeparsons commented on modules/custom/az_publication/az_publication.info.yml:
```diff
@@ -0,0 +1,31 @@
+name: Quickstart Publication (Experimental)
+description: 'Provides Publication content type'
+core_version_requirement: ^9
```
```suggestion
core_version_requirement: ^9 || ^10
```
